File: java-code/src/main/java/refactoring/basics/combine_functions_into_class/CombineFunctionsIntoClass_Before.java
package refactoring.basics.combine_functions_into_class;

public class CombineFunctionsIntoClass_Before {

  /**
  All the code using fields of reading can be moved to reading class so that we dont
   have to inject reading fields into the methods
 */
  public void calculate() {
    var aReading = acquireReading();
    var baseCharge = baseRate(aReading.getMonth(), aReading.getYear()) * aReading.getQuantity();
    var taxableCharge = Math.max(0, baseCharge - taxThreshold(aReading.getYear()));

  }

  private int taxThreshold(Integer year) {
    return 0;
  }

  private int baseRate(Integer month, Integer year) {
    return 0;
  }

  private void baseRate() {

  }

  private Reading acquireReading() {
    return null;
  }
}

File: go-code/basics/combine_functions_into_class/combine_functions_into_class_before.go
package combine_functions_into_class

/*
All the code using fields of reading can be moved to reading struct func
*/
func calculate() int {
	reading := acquireReading()
	baseCharge := baseRate(reading.Month, reading.Year) * reading.Quantity

	return Max(0, baseCharge-taxThreshold(reading.Year))

}

/**
 * reading = {customer: "ivan", quantity: 10, month: 5, year: 2017};
 */
type ReadingBefore struct {
	Month    int
	Year     int
	Quantity int
}

func taxThreshold(year int) int {
	return 0
}

func baseRate(month int, year int) int {
	return 0
}

func acquireReading() *ReadingBefore {
	return new(ReadingBefore)
}

func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
